Test 2 Liveness

Ayiga opened the pull request:
<!-- Closes #<ISSUE_NUMBER> -->
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->

Introduces configurations that allow the interception and handling of requests to the `espresso-dev-node` so that we can better simulate temporary outages, or errant behavior from the Espresso System.

Introduces a test that is meant to ensure that we retain fast confirmation guarantees in the event of Espresso network
degradation. 

<!-- ### This PR does not: -->
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->

<!-- ### Key places to review: -->
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1209731736667118
  - https://app.asana.com/0/0/1209467161957300

Ayiga committed:
Add e2e option to allow intercept

In order to effectively test the liveness requirement of our
integration, we want to ensure that we can withstand temporary
outages / downtime and still provide the same finality guarantees.

The ideal way to perform these sort of checks would be to turn
off subsystems within the Espresso environment itself.  However,
this is not currently possible.  The next best thing we can do is
to intercept the request going to Espresso, and handle them
directly or indirectly in order to simulate the desired behavior.

In order to do this we need to be able to intercept the http
requests being made against the Espresso System.

This change allows us to configure an option to intercept
requests going to Espresso's Query Service endpoint, so
that we might decide how to handle them in order to
simulate the desired behavior.

Ayiga committed:
Fix proxy request not setting headers

Ayiga committed:
Update intercept to simulate failure with rng

Update test to trigger 10% failure rate on submit

Ayiga committed:
Refactor decision making behavior of intercept

The intercept's interface and handling is a little confusing and unwieldly.
I have made efforts to refactor it so that al ot of the specific details need
not be worried about directly, while still attempting to ensure that the
caller has the ability to override the specific behavior progression he/she
may desire.

The result is more flexibility with less specific configuration required.

Ayiga committed:
Modify the liveness test to speed it up

The liveness test base case runs pretty quickly, but it can be sped up if
we don't run a full L1 -> L2 flow.  If we instead just wait for the receipt on
the L2 Sequencer, then verify that it appears on a L2 Verifier later, we
can end up with a much quicker to run test.

This separates these pieces out for speed

dailinsubjam reviewed (commented):
So this PR will not include the work-around "modified the loop be bounded, to retry no more than 10 times"?

Ayiga commented:
No, that would be a different change.  I figure we can all come together to decide what makes sense in that submission area.

philippecamacho commented on espresso/environment/2_espresso_liveness_test.go:
```diff
@@ -0,0 +1,120 @@
+			receipts = append(receipts, receipt)
+		}
+
+		// Let's verify that all of our transactions came through successfully
```
This seems redundant as [SendL2TxWithID](https://github.com/EspressoSystems/optimism-espresso-integration/blob/26cd5efe8acb2afd099dceb5c6e7c36cf5820da5/op-e2e/system/helpers/tx_helper.go#L91) already checks if the receipt is `types.ReceiptStatusSuccessful`.

Ayiga replied:
This is true, but it only verified them on the l2Seq, as no validator nodes were passed to SendL2TxWithID. This second section, is verifying that they exist on the l2Verif as well, which would indicate that the state is derivable without reading from the l2Seq.

philippecamacho replied:
I see, so this is how we should send the transactions without waiting for the "long" confirmation. Thanks.

philippecamacho commented on espresso/environment/query_service_intercept.go:
```diff
@@ -0,0 +1,387 @@
+
+// ServeHTTP implements http.Handler
+func (fakeSubmitTransactionSuccess) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// We could do a lot of effort to validate the request, and return an
```
```suggestion
	// We could do a lot of effort to validate the request, and return a
```

Ayiga replied:
Corrected in ee150d3c745734b16d2b880c74f38d940fd6cce8.

philippecamacho commented on espresso/environment/query_service_intercept.go:
```diff
@@ -0,0 +1,387 @@
+	}
+}
+
+// SetHTTPClient sets the HTTP client for the EspressoDevNodeIntercept.
```
Looks like this function is not used.

Ayiga replied:
I tis not, currently.  It was provided mainly for convenience for others to override / utilize as needed.  But if no value is seen in it, it can be removed.

philippecamacho replied:
I see, no harm in keeping it.

philippecamacho commented on espresso/environment/query_service_intercept.go:
```diff
@@ -0,0 +1,387 @@
+
+	// We should probably validate the request contents and format here, but
+	// we will just assume the settings.
+	defer r.Body.Close()
```
Should not we also simulate a query that just fails in addition to returning a randomly generated hash value?

Ayiga replied:
We **COULD** also do this.  The acceptance criteria for the test stated to turn the `builder` on and off randomly, which would not result in any noticeable affect via the query endpoint.  Though, the original criteria also stated to turn off random espresso nodes, which **could** exhibit this kind of behavior.

philippecamacho replied:
Would it be a lot of work?

Ayiga replied:
I could simulate a load balancer returning a status code indicating that the service is unreachable.  I won't be able to simulate an error connecting to the service itself.  That would require intercepting the http client itself, which we don't have the ability to do with the current `espresso-network-go` library.

philippecamacho replied:
> I could simulate a load balancer returning a status code indicating that the service is unreachable. I won't be able to simulate an error connecting to the service itself. 

Why can't the fake http server return some error status code from time to time (instead of return a random hash)? Or is this what you mean?

> That would require intercepting the http client itself, which we don't have the ability to do with the current espresso-network-go library.

Right no need to do that I think.

Ayiga replied:
> Why can't the fake http server return some error status code from time to time (instead of return a random hash)? Or is this what you mean?

This is exactly what I mean.  I would return a [503](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/503) to simulate the service being unavailable. It doesn't really matter which status code is returned as long as it is an error status code.  I just figured Service Unavailable would be the easiest thing to concretely reason about as opposed to something like [I'm a teapot](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/418).

Ayiga replied:
Oh! And other status codes, specifically those in the `4xx` range might have an assumed serialized response body of `application/json` which would be more effort to represent effectively.

philippecamacho replied:
Awesome, let us go for the 503 error then. Thanks!

Ayiga replied:
Done in ee150d3c745734b16d2b880c74f38d940fd6cce8.

philippecamacho replied:
That was fast!

Ayiga committed:
Update liveness test to simulate unreachable service

The acceptance criteria for the test states to only turn the builder off
and on which would only present itself as some submitted transactions
not being returned / confirmed by espresso.  However, the original
criteria also states to turn off random components of the espresso
network, which could present themselves as the services being
unavailable temporarily.

In order to help facilitate this design, a new behavior and handler have
been added to the Intercept to simulate an unreachable service.

philippecamacho reviewed (approved):
Great work!
